feat: Introduce LVM resource management

SK1Y101 opened the pull request:
## Description of changes

Add resource for maas LVM storage layouts, and their associated CRUD operations:

```terraform
resource "maas_logical_volume" "test" {
  fs_type 		 = "etx4"
  machine 		 = data.maas_machine.machine.id
  mount_point            = "/var"
  mount_options          = "defaults,noatime"
  name 			 = "test LVM"
  size_gigabytes         = 70
  volume_group 	         = maas_volume_group.test.id
}
```

will create an LVM on top of volume_group.test

## Issue or ticket link (if applicable)

Requires #293 to land first.

Resolves: #215, adding LVM support 

## Checklist

- [x] I have written a PR title that follows the advice in DEVELOPMENT.md with the title format `type(scope): title`
- [x] I have written a description of the changes in the PR that is clear and concise.
- [x] I have updated the documentation to reflect the changes.
- [x] I have added or updated the tests to reflect the changes.
- [x] I have run the unit tests and they pass.
- [x] I have run the acceptance tests and they pass.

skatsaounis reviewed (changes requested):
The PR looks good so far. Let me compose a list of things that are missing:

- we need to support mounting the logical volume like we do with physical block devices
- we need to support update function
- what is the purpose of adding the suffix `lvm` everywhere? I think that we have to remove it as logical volume (LV) is a key component of LVM. If we had to introduce it, then prefix makes more sense, but at the same time we would have to add it as prefix to volume group (VG) as well.

SK1Y101 commented:
>     * we need to support mounting the logical volume like we do with physical block devices
Should this be an operation, or a property like `mount_point` on the LVM

>     * we need to support update function
The API, and as far as i can tell MAAS itself, doesn't really have a notion of updating an LVM. it's treated like a blockdevice, so i assume we can use BD operations, but what would you actually update?

>     * what is the purpose of adding the suffix `lvm` everywhere? I think that we have to remove it as logical volume (LV) is a key component of LVM. If we had to introduce it, then prefix makes more sense, but at the same time we would have to add it as prefix to volume group (VG) as well.
Yeah fair, updated and removed

tobiasdemendonca commented on maas/resource_maas_logical_volume.go:
```diff
@@ -0,0 +1,215 @@
+			},
+			"machine": {
+				Type:        schema.TypeString,
+				Required:    true,
```
Could be debatable but looking at `maas_block_device` this I would perhaps expect this to be the case if the idea of updating to a new machine doesn't make sense
```suggestion
				Required:    true,
				ForceNew:    true,
```

tobiasdemendonca replied:
This seems to reflect in your update function I think

SK1Y101 replied:
Hmm yeah, that's a good point

tobiasdemendonca commented on maas/resource_maas_logical_volume.go:
```diff
@@ -0,0 +1,215 @@
+	}
+
+	params := entity.BlockDeviceParams{
+		Name: d.Get("name").(string),
```
Either you're missing "size" here or I think it should be `ForceNew`

SK1Y101 replied:
I don't think you can change the size of an LVM, maybe shrink it, but definitely can't grow it beyond the volume groups size.

tobiasdemendonca commented on maas/resource_maas_logical_volume_test.go:
```diff
@@ -0,0 +1,176 @@
+
+	machine := os.Getenv("TF_ACC_BLOCK_DEVICE_MACHINE")
+
+	size := 70
```
Nit: Not familiar with how big these things normally are but should we be making this smaller? I know we've got a lot of space but this seems kinda wildly big imo

SK1Y101 replied:
Oh right yeah, I was just throwing numbers around. will decrease this by an order of magnitude

tobiasdemendonca commented on maas/resource_maas_logical_volume_test.go:
```diff
@@ -0,0 +1,176 @@
+
+resource "maas_block_device" "lvm_bd1" {
+  machine        = data.maas_machine.machine.id
+  name           = "lvm_bd1"
```
if we can have block devices of the same name this is fine otherwise i'd suggest adding a random prefix

tobiasdemendonca replied:
Same for other names

tobiasdemendonca commented on maas/resource_maas_logical_volume_test.go:
```diff
@@ -0,0 +1,176 @@
+resource "maas_block_device" "lvm_bd1" {
+  machine        = data.maas_machine.machine.id
+  name           = "lvm_bd1"
+  size_gigabytes = 25
```
Similar suggestion for this whole config on if we need to allocate this amount of space for this test

tobiasdemendonca commented on maas/resource_maas_logical_volume_test.go:
```diff
@@ -0,0 +1,176 @@
+			return fmt.Errorf("error getting the logical volume: %s", err)
+		}
+
+		*logicalVolume = *gotLogicalVolume
```
Do we need this at all, I've seen this done in other tests but I don't see the point if it's not being used later, even defining this variable at the start of the tests
```suggestion
```

tobiasdemendonca reviewed (changes requested):
Your PR reads super nicely, I'm not that familiar with LVMs but can get the idea from just the clear function names with few comments, lovely stuff. 

A few comments here and there

SK1Y101 commented on docs/resources/logical_volume.md:
````diff
@@ -0,0 +1,74 @@
+
+```terraform
+resource "maas_block_device" "vdb1" {
+  machine        = maas_machine.virsh_vm2.id
````
Yep, excellent point, will add that

skatsaounis commented:
Blocked by: https://github.com/canonical/gomaasclient/issues/123

skatsaounis commented on maas/resource_maas_logical_volume_test.go:
```diff
@@ -0,0 +1,172 @@
+  fs_type 		 = %q
+  machine 		 = data.maas_machine.machine.id
+  name 			 = %q
+  volume_group 	 = maas_volume_group.lvm_vg.id
```
I think we need `terraform fmt` here

skatsaounis commented on maas/resource_maas_logical_volume.go:
```diff
@@ -0,0 +1,219 @@
+			"size_gigabytes": {
+				Type:        schema.TypeInt,
+				Required:    true,
+				ForceNew:    true,
```
This `ForceNew` should be removed. Most probably it is a remnant from previous commits when the gomaasclient in use could not properly set the size

skatsaounis reviewed (changes requested):
Some minor changes and we are good to go! Plus, please update to latest master

tobiasdemendonca commented on maas/resource_maas_logical_volume.go:
```diff
@@ -0,0 +1,218 @@
+			"mount_options": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The options used for the logical volume mount.",
```
Just for my understanding is this comma separated mount options?

SK1Y101 replied:
Yes, i believe so, I'll double check and then specify

tobiasdemendonca reviewed (commented):
Minor nit

SK1Y101 commented on maas/resource_maas_logical_volume.go:
```diff
@@ -0,0 +1,219 @@
+	}
+}
+
+func resourceMAASLogicalVolumeCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```
Oh, fair, yeah they do! thanks

tobiasdemendonca reviewed (approved):
LGTM
